### DISCUSSION TITLE
Reading String Array types from Collect_Unique query output

### DISCUSSION THREAD

**Discussion by ShivaniSK** [finos/TimeBase-CE#227] [Q&A] [answered]
Hello,
We are executing below query using COLLECT_UNIQUE which groups unique string values and gives an array column as output.
However we are unable to read that array in java class as well as in timebase shell.
Query: `select COLLECT_UNIQUE{} (column1) as columnArray from table_name where SYMBOL == 'ABC' GROUP BY timestamp`
Output:
`
>_,TIMESTAMP,SYMBOL,TYPE,columnArray
0,2022-08-26 00:00:00,ABC,**ARRAY**
1,2022-08-26 00:01:00,ABC,**ARRAY**
2,2022-08-26 00:02:00,ABC,**ARRAY**
3,2022-08-26 00:03:00,ABC,**ARRAY**
4,2022-08-26 00:06:00,ABC,**ARRAY**
5,2022-08-26 00:09:00,ABC,**ARRAY**
`

The expected output for ARRAY was [aaa, aad] .. an array of strings since column1 holds String values.
As shown above the timebase shell shows "ARRAY" as output instead of showing actual array values.

Also we are unable to read this string array in a java model class.

We tried using below field types in Java class

protected ObjectArrayList<String> columnArray = null;
protected ObjectArrayList<Object> columnArray = null;

None of the above is able to read the array output from above query.

Can you please help us out.

Regards,
Shivani

**Comment by ShivaniSK** [comment]
Hi @alex-karpovich ,
Do we have any update on this?

**Comment by ShivaniSK** [comment]
Hello @alex-karpovich  do we have any update on this?

**Comment by Romkisel** [comment, marked as answer]
Hi, ShivaniSK

To be able to print messages as json in timebase shell, call `set printJson true` before query.

To be able to read array of strings into custom Java class you should map query output type to your java class, use the following steps:

1) Create custom class:

```
package sample;

import com.epam.deltix.timebase.messages.InstrumentMessage;
import com.epam.deltix.timebase.messages.SchemaArrayType;
import com.epam.deltix.timebase.messages.SchemaElement;
import com.epam.deltix.util.collections.generated.ObjectArrayList;

public class TestMessage extends InstrumentMessage {

    private ObjectArrayList<CharSequence> columnArray;

    @SchemaElement
    @SchemaArrayType
    public ObjectArrayList<CharSequence> getColumnArray() {
        return columnArray;
    }

    public void setColumnArray(ObjectArrayList<CharSequence> columnArray) {
        this.columnArray = columnArray;
    }
}
```

2) Use `TYPE` QQL keyword to specify output message type:

```
select COLLECT_UNIQUE{} (column1) as columnArray TYPE "sample.TestMessage" from table_name where SYMBOL == 'ABC' GROUP BY timestamp
```

3) Execute query and read result:

```
InstrumentMessageSource cursor = db.executeQuery(
    "select COLLECT_UNIQUE{} (column1) as columnArray TYPE \"sample.TestMessage\" from table_name where SYMBOL == 'ABC' GROUP BY timestamp"
);
if (cursor.next()) {
    InstrumentMessage message = cursor.getMessage();
    if (message instanceof TestMessage) {
       //...
    }
}
```

**Reply by ShivaniSK** [reply]
Thanks @alex-karpovich . This helps :)
